File: apps/web/lib/graph.ts
// The experiment graph, as the surface reads it.
//
// The ladder model this replaces described the campaign as an ordered
// sequence of steps, each a grid of jobs. That shape could not say what a
// step had established, only how many cells had finished, and it kept
// reporting a full grid after the evidence behind it was deleted.
//
// The graph model asks a different question of the record. A NODE is one
// decision over a field or a group of inseparable fields. The EDGE into a
// node carries a STRENGTH saying what the record can support about that
// decision, and the five values are exhaustive and mutually exclusive:
// a comparison that separated, a comparison that ran and did not, a value
// nobody ever chose, a comparison that could not have resolved anything,
// and a level that cannot be produced at all.
//
// Nothing here derives, filters or repairs the payload. Every number on
// the page is the number the endpoint returned, so a reader chasing a
// surprising cell lands in one query rather than in a client-side pipeline.

import { ApiError, baseUrl } from "@/lib/api";

/**
 * What the record supports about one decision.
 *
 * - `measured`   a declared comparison separated against its floor.
 * - `chosen`     the comparison ran with power, did not separate, and the
 *                value was picked and recorded.
 * - `inherited`  nobody ever decided; the value is the one it always was.
 * - `unpowered`  the comparison could not have resolved anything. Known
 *                from the design, before the comparison ran.
 * - `blocked`    a level cannot be produced because its artifact has no
 *                producer.
 */
export type EdgeStrength =
  | "measured"
  | "chosen"
  | "inherited"
  | "unpowered"
  | "blocked";

/** Declaration order, which is also pipeline order on the page. */
export const EDGE_STRENGTHS: EdgeStrength[] = [
  "measured",
  "chosen",
  "inherited",
  "unpowered",
  "blocked",
];

export function isEdgeStrength(v: unknown): v is EdgeStrength {
  return typeof v === "string" && (EDGE_STRENGTHS as string[]).includes(v);
}

/**
 * The frame: which window, scored against what, attributable or not.
 *
 * `sealed_rows` / `unsealed_rows` are the load-bearing pair. A result row
 * that does not name the frame it was produced under cannot be attributed
 * to one, and a number that cannot be attributed cannot be compared. The
 * page shows the unsealed count before it shows any score.
 */
/** When a window opened, when it closed, and how long it ran. */
export type WindowSpan = {
  from: string;
  to: string;
  days: number;
  months: number;
};

/** One dated release on the frame's axis, and its part in the frame. */
export type TimelineMark = {
  kind: "annotation_set" | "ontology_snapshot";
  label: string;
  date: string;
  role: "window_start" | "window_end" | "pivot" | "inside" | "before" | "beyond";
  in_window: boolean;
  is_pivot: boolean;
};

/** Every dated release the record holds, laid out against the window. */
export type GraphTimeline = {
  window: { from: string | null; to: string | null };
  marks: TimelineMark[];
};

export type GraphFrame = {
  /**
   * The dates behind the release numbers.
   *
   * 220 and 227 name two files and date nothing. How long a window ran bounds
   * how much annotation could accumulate in it, which is the first thing asked
   * of a temporal benchmark and the last thing a bare release number answers.
   */
  window_span: WindowSpan | null;
  declared: boolean;
  evaluation_set_id: string | null;
  /** Rendered verbatim; the endpoint formats it, e.g. `220->227`. */
  window: string | null;
  window_role: string | null;
  mode: string | null;
  pivot_snapshot: { id: string; version: string } | null;
  information_accretion_set: { id: string; regime: string; sha256: string } | null;
  query_set: { id: string; name: string; entries: number } | null;
  sealed_rows: number;
  unsealed_rows: number;
};

/** One decision over a field or a group of inseparable fields. */
/** One field of a node and the value it stands at. */
export type HeldValue = {
  field: string;
  value: string;
};

export type GraphNode = {
  key: string;
  title: string;
  /** Position in the pipeline, 1-based, ascending. */
  stage: number;
  question: string;
  strength: EdgeStrength;
  levels_instantiated: number;
  levels_available: number;
  varying_fields: string[];
  constant_fields: string[];
  /**
   * What the node currently stands at, field by field.
   *
   * A strength says how firmly a decision is held and nothing about what was
   * decided. Without the value, an inherited default nobody chose reads
   * exactly like a deliberate choice nobody measured.
   */
  held: HeldValue[];
  /** Why this node cannot answer. Null when it can. */
  blocked_reason: string | null;
  results: number;
};

/** One level's number on one panel. */
export type PanelResult = {
  level: string;
  f_micro_w: number | null;
  tau: number | null;
};

/**
 * One of the nine regions: a knowledge category by an aspect.
 *
 * Panels are never summed, averaged or otherwise collapsed. Cardinality is
 * a vector over the nine, not a scalar, and a single headline over all of
 * them is a claim the model does not license.
 */
export type GraphPanel = {
  category: string;
  aspect: string;
  units: number | null;
  /**
   * The smallest difference this panel could resolve, from its population
   * alone, at the cheapest contrast class measured. A gap smaller than this
   * is not a small result; it is no result, and the panel says so before
   * anything is run.
   */
  detectable_effect: number | null;
  results: PanelResult[];
};

/** A level that cannot be produced, and what would unblock it. */
export type GraphBlocked = {
  node: string;
  what: string;
  why: string;
  precondition: string;
};

export type GraphResponse = {
  frame: GraphFrame;
  nodes: GraphNode[];
  timeline: GraphTimeline | null;
  panels: GraphPanel[];
  blocked: GraphBlocked[];
};

/** Canonical row order for the panel grid. */
export const PANEL_CATEGORIES = ["NK", "LK", "PK"];
/** Canonical column order for the panel grid. */
export const PANEL_ASPECTS = ["BPO", "MFO", "CCO"];

/**
 * True when the endpoint answered and there is nothing behind the answer.
 *
 * Kept apart from the loading and error states on purpose. An instrument
 * that shows a skeleton forever cannot be told from one waiting on a slow
 * query, and both are wrong when the honest report is that no frame has
 * been declared and no level instantiated.
 */
export function isEmptyGraph(g: GraphResponse): boolean {
  if (g.frame.declared) return false;
  if (g.nodes.some((n) => n.levels_instantiated > 0 || n.results > 0)) return false;
  if (g.panels.some((p) => (p.units ?? 0) > 0 || p.results.length > 0)) return false;
  // A graph with nothing instantiated still has something to say when it names
  // what is blocked and what each blocked node is waiting for. Treating that as
  // empty hides the only actionable content an empty record carries.
  if (g.blocked.length > 0) return false;
  return true;
}

/** Every level name that appears on any panel, in a neutral order.
 *
 * Alphabetical rather than ranked. Ranking rows would need one number per
 * level across the nine panels, and there is no such number: collapsing
 * the panels to produce one is exactly what the model forbids. The reader
 * orders by a panel they pick, which is a claim about that panel only.
 */
export function panelLevels(panels: GraphPanel[]): string[] {
  const seen = new Set<string>();
  for (const p of panels) for (const r of p.results) seen.add(r.level);
  return [...seen].sort((a, b) => a.localeCompare(b));
}

/** Index the panels by `CATEGORY/ASPECT` so lookups do not rescan. */
export function panelKey(category: string, aspect: string): string {
  return `${category}/${aspect}`;
}

export function indexPanels(panels: GraphPanel[]): Map<string, GraphPanel> {
  return new Map(panels.map((p) => [panelKey(p.category, p.aspect), p]));
}

/** `level -> result` for one panel. */
export function indexPanelResults(panel: GraphPanel | undefined): Map<string, PanelResult> {
  if (!panel) return new Map();
  return new Map(panel.results.map((r) => [r.level, r]));
}

/**
 * The strongest level on one panel, and how far apart the levels are.
 *
 * Both are per-panel and stay per-panel. `spread` is max minus min inside
 * this panel, which is what a reader needs to judge whether the ordering
 * here is worth anything; it is never compared across panels on the page.
 */
export function panelSummary(
  panel: GraphPanel,
  metric: "f_micro_w" | "tau",
): { best: PanelResult; spread: number } | null {
  // Only results that actually carry the metric can be summarised. A stored
  // result missing it is not a zero, so it takes no part in the best or the
  // spread, and a panel where none carries it has no summary at all.
  const scored = panel.results.filter((r) => r[metric] != null);
  if (scored.length === 0) return null;
  let best = scored[0];
  let lo = scored[0][metric] as number;
  let hi = scored[0][metric] as number;
  for (const r of scored) {
    const v = r[metric] as number;
    if (v > (best[metric] as number)) best = r;
    if (v < lo) lo = v;
    if (v > hi) hi = v;
  }
  return { best, spread: hi - lo };
}

/**
 * Fetch the graph.
 *
 * `async` is load-bearing: `baseUrl()` throws synchronously when
 * NEXT_PUBLIC_API_URL is unset, and in a non-async function that throw
 * escapes before a promise exists, so a caller's `.catch()` never
 * attaches. Making it async puts a missing base URL, a bad status and a
 * wrong body shape in the one channel the caller already handles.
 */
export async function getGraph(): Promise<GraphResponse> {
  const path = "/graph";
  let res: Response;
  try {
    // Never cached. This is the surface an operator refreshes while a run
    // seals its rows, and a stale "8 unsealed" beside a sealed board is
    // worse than a slower answer.
    res = await fetch(`${baseUrl()}${path}`, { cache: "no-store" });
  } catch (e) {
    const msg = e instanceof Error ? e.message : String(e);
    throw new ApiError("network", 0, path, msg);
  }
  if (!res.ok) {
    const body = await res.text();
    const msg = body.trimStart().startsWith("<") || body.trim() === ""
      ? `HTTP ${res.status} ${res.statusText}`
      : body;
    throw new ApiError(
      res.status === 401 ? "unauthorized" : res.status === 403 ? "forbidden" : "http",
      res.status,
      path,
      msg,
    );
  }
  const body: unknown = await res.json();
  // A 200 is not an answer. The e2e mock replies 200 with `[]` to every
  // route it does not know; casting that to the response type would put a
  // TypeError outside the promise chain, where the caller's catch cannot
  // see it. Reject here so a wrong shape arrives through the same channel
  // as a wrong status.
  if (!isGraphResponse(body)) {
    throw new ApiError("http", res.status, path, "graph: response does not have the graph shape");
  }
  return body;
}

function isGraphResponse(body: unknown): body is GraphResponse {
  if (typeof body !== "object" || body === null) return false;
  const b = body as Partial<GraphResponse>;
  if (typeof b.frame !== "object" || b.frame === null) return false;
  if (typeof b.frame.declared !== "boolean") return false;
  return Array.isArray(b.nodes) && Array.isArray(b.panels) && Array.isArray(b.blocked);
}

File: protea/api/routers/graph.py
"""The experiment graph: which decisions have been taken, and on what evidence.

WHAT THIS REPLACES. The retired ladder surface described the campaign as an
ordered sequence of steps, each one a grid of jobs, and counted the grid from
the job table. That table is append-only and outlived a wipe that removed the
evaluation results, so the count went on reporting a complete measurement with
no surviving evidence behind it. Nothing here counts a job. Every number is read
off an artifact that still exists, and a node with no artifact says so rather
than borrowing a number from the queue.

THE MODEL. A FLOW is an annotation source plus a mechanism that propagates it.
Running the same mechanism over a different representation is not a second flow,
it is the same flow configured differently. A NODE is one decision over a field,
or over a group of fields that cannot be decided apart. The nodes are, in
pipeline order: frame, substrate, bank, retriever, generator, scoring, features,
re-ranking, combination, routing. The EDGE into a node says how firmly that
decision is held and takes one of five values:

``measured``
    a declared comparison separated against its floor.
``chosen``
    the comparison ran with power and did not separate; a level was selected and
    recorded.
``inherited``
    nobody ever decided. The value is the one it has always been.
``unpowered``
    the comparison could not have resolved anything. Settled from the shape of
    the comparison, before any metric is read.
``blocked``
    a level cannot be produced at all, because its artifact has no producer.

A PANEL is one of nine regions, a knowledge category (NK, LK, PK) crossed with
an aspect (BPO, MFO, CCO). Panels are never pooled and never summed: the
cardinality of this record is a vector over the nine, not a scalar.

WHY A SINGLE LEVEL IS NEVER A MEASUREMENT. One instantiated level means no
contrast existed, so no separation can have happened whatever the numbers look
like. Such a node is ``inherited`` when nothing fixed the value and ``chosen``
when the frame's own definition fixed it. Two or more levels with fewer than two
of them scored is ``unpowered``: a contrast needs two scored levels, and that is
knowable before reading a single metric. Two or more scored levels reach
``measured`` only when a floor is declared for the comparison and some level
clears it on every panel that carries both. No table in this schema has a column
for a floor, so the one place a floor can be declared is an ``experiment_run``
whose ``config`` names ``graph_node`` and ``floor``. Until such a row exists no
node here can be a measurement, and that is the honest report rather than a
defect in the endpoint.

WHERE THE PANEL POPULATIONS COME FROM. Not from the ground-truth parquet, which
this endpoint never opens, and not from ``evaluation_set.stats``, which counts
proteins per knowledge category across all three aspects at once and so cannot
answer a per-panel question at all. They are reconciled out of the stored
metrics. ``n_proteins`` on a panel is the count of proteins holding a prediction
at the threshold that run settled on, and ``coverage_at_tau`` is that same count
over the panel's ground-truth cohort, so the cohort is the ratio of the two.
Both are rounded when persisted, so each scored result pins the cohort to a
short interval rather than to a number, and intersecting the intervals of every
result that scored the panel collapses it. When the intervals do not overlap the
endpoint reports null, never a rounded guess and never a zero.

NOTHING IS WRITTEN. Every statement behind this is a SELECT, and they all live
in ``_graph_reads``. ``session_scope`` is deliberately not used here because it
commits on exit, and a reporting surface has no business opening a write path.
"""

from __future__ import annotations

from collections.abc import Mapping
from typing import Any

from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session, sessionmaker

from protea.api.deps import get_session_factory, get_settings
from protea.api.routers._graph_edges import BLOCKED, Built
from protea.api.routers._graph_nodes import (
    _bank_node,
    _combination_node,
    _features_node,
    _frame_node,
    _generator_node,
    _reranking_node,
    _retriever_node,
    _routing_node,
    _scoring_node,
    _substrate_node,
    _window_span,
)
from protea.api.routers._graph_panels import (
    build_panels,
    contrast_floors,
    panel_units_from_groundtruth,
)
from protea.api.routers._graph_reads import read_pivot_aspects, read_record
from protea.api.routers._graph_representations import build_representations
from protea.infrastructure.settings import Settings
from protea.infrastructure.storage.factory import (
    ArtifactStoreUnavailable,
    get_artifact_store,
)

router = APIRouter(prefix="/graph", tags=["graph"])

# The edge vocabulary, closed at five. Spelled once so a typo in a builder is an
# import-time NameError rather than a word nobody downstream recognises.

# ── The frame ─────────────────────────────────────────────────────────────────


def build_frame(record: dict[str, list[dict[str, Any]]]) -> dict[str, Any]:
    """The frame every number below is read in, and whether it is declared.

    ``declared`` is not "a frame exists". It is true only when the record leaves
    a reader nothing to guess: exactly one evaluation set, so there is a single
    window under measurement; both ends of that window resolve to a release; a
    pivot snapshot, an accretion table and a query set all resolve; and no
    published result is unsealed. A result is sealed when its ``frame`` column
    says which scoring frame it lives in. Today every one of them is null, which
    is why the block below can be fully populated and still report false: the
    frame is recoverable, but no published number states it.
    """
    heads = record["evaluation_sets"]
    head = heads[0] if heads else None
    accretion = next((a for a in record["accretion"] if a["in_use"]), None)
    query_set = next((q for q in record["query_sets"] if q["in_use"]), None)
    results = record["results"]
    sealed = sum(1 for r in results if r["frame"])
    unsealed = len(results) - sealed
    ends = (head["window_from"], head["window_to"]) if head else (None, None)
    window = f"{ends[0]}->{ends[1]}" if all(ends) else None
    # The release numbers name two files and date nothing. Fourteen months
    # separate 220 from 227, and how long a window ran is the first thing asked
    # of a temporal benchmark: it bounds how much annotation could accumulate,
    # and therefore what any panel drawn from it could possibly resolve.
    span = _window_span(head) if head else None
    pivot = (
        {"id": head["pivot_snapshot_id"], "version": head["pivot_version"]}
        if head and head["pivot_snapshot_id"]
        else None
    )
    return {
        "declared": bool(
            head is not None
            and len(heads) == 1
            and window
            and pivot
            and accretion
            and query_set
            and unsealed == 0
        ),
        "evaluation_set_id": head["id"] if head else None,
        "window": window,
        "window_span": span,
        "window_role": head["window_role"] if head else None,
        "mode": head["mode"] if head else None,
        "pivot_snapshot": pivot,
        "information_accretion_set": (
            {"id": accretion["id"], "regime": accretion["regime"], "sha256": accretion["sha256"]}
            if accretion
            else None
        ),
        "query_set": (
            {"id": query_set["id"], "name": query_set["name"], "entries": query_set["entries"]}
            if query_set
            else None
        ),
        "sealed_rows": sealed,
        "unsealed_rows": unsealed,
    }


# ── Assembly ──────────────────────────────────────────────────────────────────


def build_timeline(
    marks: list[dict[str, Any]], head: dict[str, Any] | None
) -> dict[str, Any] | None:
    """The frame laid out on a date axis, with each release's part in it named.

    A table can say the window runs from 220 to 227 and the pivot is the July
    graph. Only an axis shows what that arrangement means: the pivot sits inside
    the window it reconciles rather than at either end, an ontology contemporary
    with the window's opening exists and was not the one chosen, and a release
    beyond the closing end is the cohort nobody is allowed to look at yet.

    Every role is derived from dates the record already holds. Nothing is
    positioned by assumption, and a release the record cannot date does not
    appear at all rather than appearing at a guessed end.
    """
    if not head or not marks:
        return None
    start, end = head.get("window_from_date"), head.get("window_to_date")
    pivot_version = head.get("pivot_version")
    out: list[dict[str, Any]] = []
    for mark in marks:
        when = mark.get("date")
        if not when:
            continue
        inside = bool(start and end and start <= when <= end)
        if when == start:
            role = "window_start"
        elif when == end:
            role = "window_end"
        elif mark["version"] == pivot_version:
            role = "pivot"
        elif start and when < start:
            role = "before"
        elif end and when > end:
            role = "beyond"
        else:
            role = "inside"
        out.append(
            {
                "kind": mark["kind"],
                "label": mark["label"],
                "date": when,
                "role": role,
                "in_window": inside,
                "is_pivot": mark["version"] == pivot_version,
            }
        )
    return {
        "window": {"from": start, "to": end},
        "marks": sorted(out, key=lambda m: m["date"]),
    }


def build_graph(
    record: dict[str, list[dict[str, Any]]],
    units: Mapping[tuple[str, str], int] | None = None,
) -> dict[str, Any]:
    """Turn the rows into the graph. A pure function of what was read.

    The ``blocked`` list is derived from the nodes rather than compiled beside
    them, so a node cannot report itself blocked and then be missing from the
    list, and cannot appear in the list without the reason its own entry
    carries. The two say the same thing because they are built from one source.
    """
    head = record["evaluation_sets"][0] if record["evaluation_sets"] else None
    floors = {f["node"]: f["floor"] for f in record["floors"] if f["node"] and f["floor"]}
    # EVERY builder gets the floors, not one of them. Until 2026-09-02 this list
    # handed the dict to _scoring_node alone, so nine of the ten nodes could not
    # see a declared floor at all and `strength_of` returned CHOSEN for them
    # whatever anyone declared. The vocabulary said `measured` was reachable and
    # for nine nodes it was not, which is a surface asserting a state it cannot
    # produce -- the same defect this endpoint exists to end.
    built: list[Built] = [
        _frame_node(record, head, floors),
        _substrate_node(record, floors),
        _bank_node(record, head, floors),
        _retriever_node(record, floors),
        _generator_node(record, floors),
        _scoring_node(record, floors),
        _features_node(record, floors),
        _reranking_node(record, floors),
        _combination_node(record, floors),
        _routing_node(record, floors),
    ]
    return {
        "frame": build_frame(record),
        "nodes": [node for node, _, _ in built],
        # The Substrate node's denominator, expanded. It is beside the nodes
        # rather than inside one because every other node's levels are named by
        # their fields and this node's are not: a representation is a model, a
        # layer, a pooling and a coverage over the corpus, and none of that fits
        # in a field list.
        "representations": build_representations(record),
        "panels": build_panels(record["panels"], units),
        # The floors the panels and every finer cell are read against. A
        # constant of the design rather than a fact about this record, which is
        # why it is served with the record instead of being restated by every
        # surface that has to mark a cell as too thin to decide anything on.
        "floors": contrast_floors(),
        "timeline": build_timeline(
            record.get("timeline", []),
            record["evaluation_sets"][0] if record["evaluation_sets"] else None,
        ),
        "blocked": [
            {
                "node": node["key"],
                "what": what,
                "why": node["blocked_reason"],
                "precondition": precondition,
            }
            for node, what, precondition in built
            if node["strength"] == BLOCKED
        ],
    }


@router.get("")
def get_graph(
    factory: sessionmaker[Session] = Depends(get_session_factory),
    settings: Settings = Depends(get_settings),
) -> dict[str, Any]:
    """The experiment graph as the record currently supports it.

    Reads and returns. The session is opened directly rather than through
    ``session_scope``, which commits on exit; there is nothing to commit and a
    reporting surface should not hold a transaction open that could.
    """
    with factory() as session:
        record = read_record(session)
        aspect_of_term = read_pivot_aspects(session, record)
    units = _counted_units(record, aspect_of_term, settings)
    return build_graph(record, units)


def _counted_units(
    record: dict[str, list[dict[str, Any]]],
    aspect_of_term: dict[str, str],
    settings: Settings,
) -> Mapping[tuple[str, str], int] | None:
    """Count the panel populations from the window's stored ground truth.

    Returns ``None`` when the artefact cannot be read, so every panel reports an
    absent population rather than one standing in for it. That is the whole
    point: the population is the one quantity on this page that must not be
    inferred, because a plausible wrong number here is indistinguishable from a
    right one and silently rescales every panel that reads it.
    """
    head = record["evaluation_sets"][0] if record["evaluation_sets"] else None
    uri = (head or {}).get("groundtruth_uri")
    if not uri or not aspect_of_term:
        return None
    try:
        store = get_artifact_store(settings)
        payload = store.get(uri.split("/", 3)[3] if uri.startswith("s3://") else uri)
    except (ArtifactStoreUnavailable, OSError, ValueError, KeyError):
        return None
    return panel_units_from_groundtruth(payload, aspect_of_term)
